### PR TITLE
Fix handling of multi-column constraints

### PR DESCRIPTION
**Pull request by ronocod**
### Problem

When inserting into a table an update is tried first to provide upsert functionality. In the WHERE clause of the update, only the first column of any multi-column constraint is used.

For this table:
```sql
CREATE TABLE IF NOT EXISTS 'superhero' (
        _id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
        name TEXT NOT NULL,
        power TEXT NOT NULL,
        comic_company TEXT NOT NULL,
        UNIQUE (name, comic_company) ON CONFLICT REPLACE
);
```
only one entry would be allowed for each `name` value, when really one entry should be allowed for each combination of `name` and `comic_company` values.

### Solution

Create the WHERE clause using every column in the constraint.

* Adds a `Constraint` class which wraps a list of column names
* Uses this class instead of a single `String` where needed
* Deprecates the existing methods that use `String` instead of `Constraint`

### CODE REVIEW

**Comment by blundell** [comment]
:x:  needs a test in the automation module and needs an example in one of the two demo modules

**Comment by xrigau** [comment]
Fixes https://github.com/novoda/sqlite-provider/issues/24

**Comment by ronocod** [comment]
* Added unit test for finding all columns in a multi-column constraint
* Added a multi-column constraint to the shop table in the extended demo

**Comment by blundell** [comment]
:+1: :shipit:

**Comment by blundell** [comment]
10 mins an i'll merge

**Comment by xrigau** [comment]
you forgot fixes #24

File: core/src/main/java/novoda/lib/sqliteprovider/util/Constraint.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        this.columns = columns;
+    }
+
+    public List<String> getColumns() {
+        return columns;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (xrigau) [inline core/src/main/java/novoda/lib/sqliteprovider/util/Constraint.java]: missing braces

**Comment by xrigau** [comment]
:+1:

File: core/src/main/java/novoda/lib/sqliteprovider/util/Constraint.java (hunk 1 above)
Comment (ronocod) [inline core/src/main/java/novoda/lib/sqliteprovider/util/Constraint.java]: IntelliJ auto-generated

**Comment by mr-archano** [comment]
Great job @ronocod!

**Pull request merged**